File: hello-grpc-swift/Sources/Common/Otel.swift
import Foundation
import GRPCCore
import OpenTelemetryApi
import OpenTelemetrySdk
import StdoutExporter

/// hello-grpc-swift OpenTelemetry wiring.
///
/// Mirrors the env-gate pattern used by the prior OTel PRs across
/// the other languages. opt-in via `GRPC_HELLO_OTEL=Y`.
public enum Otel {
    public static let envEnabled = "GRPC_HELLO_OTEL"

    /// Returns `true` iff `GRPC_HELLO_OTEL=Y`.
    public static var enabled: Bool {
        return ProcessInfo.processInfo.environment[envEnabled] == "Y"
    }

    private static let lock = NSLock()
    private static var _initialized = false

    /// One-time SDK setup. No-op when the env var is unset.
    public static func initOtel(_ serviceName: String) {
        guard enabled else { return }
        lock.lock()
        defer { lock.unlock() }
        guard !_initialized else { return }
        _initialized = true

        let exporter = StdoutExporter()
        let processor = SimpleSpanProcessor(spanExporter: exporter)
        let resource = DefaultResources.get().merging(
            with: Resource(attributes: [
                "service.name": AttributeValue(serviceName),
            ])
        )

        OpenTelemetry.registerTracerProvider(
            tracerProvider: TracerProviderBuilder()
                .add(spanProcessor: processor)
                .with(resource: resource)
                .build()
        )
    }

    /// Global tracer from the configured provider, or the no-op tracer.
    public static var tracer: Tracer {
        return OpenTelemetry.instance.tracerProvider.get(
            instrumentationName: "hello-grpc-swift"
        )
    }

    /// Starts a span with rpc.system/grpc attributes if OTel is enabled.
    public static func startSpan(
        _ name: String,
        kind: SpanKind = .internal
    ) -> Span? {
        guard enabled else { return nil }
        return tracer.spanBuilder(spanName: name)
            .setSpanKind(kind)
            .setAttribute(key: "rpc.system", value: "grpc")
            .setAttribute(key: "rpc.method", value: name)
            .startSpan()
    }
}

/// gRPC Swift 2.x server interceptor that emits an inbound [SpanKind.server] span
/// around each RPC.
public struct HelloServerInterceptor: ServerInterceptor, Sendable {
    public init() {}

    public func intercept<Input: Sendable, Output: Sendable>(
        request: StreamingServerRequest<Input>,
        context: ServerContext,
        next: @Sendable (
            _ request: StreamingServerRequest<Input>,
            _ context: ServerContext
        ) async throws -> StreamingServerResponse<Output>
    ) async throws -> StreamingServerResponse<Output> {
        guard let span = Otel.startSpan(context.descriptor.fullName, kind: .server)
        else {
            return try await next(request, context)
        }

        defer { span.end() }

        do {
            return try await next(request, context)
        } catch {
            span.setStatus(.error, description: String(describing: error))
            throw error
        }
    }
}

/// gRPC Swift 2.x client interceptor that emits an outbound [SpanKind.client] span
/// around each RPC.
public struct HelloClientInterceptor: ClientInterceptor, Sendable {
    public init() {}

    public func intercept<Input: Sendable, Output: Sendable>(
        request: StreamingClientRequest<Input>,
        context: ClientContext,
        next: @Sendable (
            _ request: StreamingClientRequest<Input>,
            _ context: ClientContext
        ) async throws -> StreamingClientResponse<Output>
    ) async throws -> StreamingClientResponse<Output> {
        guard let span = Otel.startSpan(context.descriptor.fullName, kind: .client)
        else {
            return try await next(request, context)
        }

        defer { span.end() }

        do {
            return try await next(request, context)
        } catch {
            span.setStatus(.error, description: String(describing: error))
            throw error
        }
    }
}

File: hello-grpc-dart/lib/common/otel.dart
// ignore_for_file: avoid_print
import 'dart:async';
import 'dart:io';

import 'package:opentelemetry_dart/opentelemetry_dart.dart';

/// hello-grpc-dart OpenTelemetry wiring.
///
/// Mirrors the env-gate pattern in the prior OTel PRs across the
/// other languages. opt-in via `GRPC_HELLO_OTEL=Y`.
///
/// Caveat: grpc-dart does not yet ship an OpenTelemetry client/
/// server interceptor at the time of writing, so this module
/// exposes a minimal tracer that future interceptor code can be
/// built around. The env var also installs a tracer provider so
/// that any in-process tracer usage (service handlers, custom
/// async spans) works against a configured global tracer. The
/// interceptor half follows in a follow-up PR.

bool _enabled = false;
bool _initialized = false;

bool otelEnabled() {
  if (_enabled) return true;
  final v = Platform.environment['GRPC_HELLO_OTEL'];
  _enabled = v == 'Y';
  return _enabled;
}

/// One-time TracerProvider + tracer install for the calling
/// process. Idempotent. No-op when otel is not enabled.
Future<void> initOtel(String serviceName) async {
  if (!otelEnabled() || _initialized) return;
  _initialized = true;

  // Process-level resource. The default global propagator + exporter
  // wiring is provided by package:opentelemetry_dart; stdout is the
  // teaching/demo posture across the other languages.
  final resource = OTelResource(
    serviceName: serviceName,
    serviceVersion: '0.1.0',
  );
  final exporter = OTelConsoleExporter();
  final provider = OTelTracerProvider(resource: resource, exporter: exporter);
  OTelTrace.register(provider, setAsGlobal: true);

  // Print to stdout so the user can see wiring took effect.
  // ignore: avoid_print
  print('[otel] hello-grpc-dart tracing enabled for $serviceName');
}
